refactor app.py with ActivedopTree class abstraction

bwaldon opened the pull request:
Introduces a new class `ActivedopTree`, with built-in methods for common tree format conversions.

bwaldon commented:
A further opportunity for refactoring: condense the two classmethods `from_cgeltree` and `from_brackettreestr` into one `from_str` method whose behavior depends on the value of ` app.config['CGELVALIDATE']`. (Currently, there are several spots in the code where we call one or the other classmethod depending on the value of this config param).

bwaldon commented:
Another possible opportunity: call the validator on `__init__` and save the output message as an object attribute.

bwaldon commented:
SUMMARY: Introduces the `ActivedopTree` class as a new abstraction in app.py. `ActivedopTree` class serves as a wrapper for a `ParentedTree` object, providing additional methods and attributes specific to activedop.

**Constructor** (`__init__` method):
Initializes the ActivedopTree object with a ParentedTree `ptree` and a list of tokens `senttok`. Canonicalizes `ptree` and creates a CGEL tree conversion with a call to `tree_process`. Optionally updates the CGEL tree terminals if `cgel_terminals` are provided.

**Attributes**:
`ptree`: The ParentedTree object with numeric indices as terminals.
`senttok`: A list of tokens corresponding to the terminals of `ptree`.
`cgel_tree`: A CGELTree object.

**Methods**:
`brackettreestr`: Returns a string representation of `ptree` in bracket notation, with an option for pretty printing (which is helpful in standard, non-CGEL mode).
`treestr`: Returns a string representation of the tree (either CGEL or bracket notation, depending on app settings).
`gtree`: Returns an HTML representation of `ptree`, with an option to add an `editable` attribute to select tree nodes using  `add_editable_attribute`.
`validate`:  Returns a string, the output of validation using the standard `validate()` function (plus `validate_cgel()`, if applicable). 
`from_str` (classmethod): create an `ActivedopTree` instance from a raw string (either brackettree format or cgeltree format, depending on app settings).

nschneid commented on app.py:
```diff
@@ -153,6 +153,59 @@ def sent_escape(sent):
+	cgel_terminals: list of CGELTree terminals.
+	brackettreestr: string representation of ptree in bracket notation, with labels consisting of a POS tag and a function tag separated by a hyphen.
+	treestr: helpful alias for string representation of the tree (CGEL or bracket notation depending on app settings).
+	gtree: html representation of ptree.
```
The method descriptions should be in triple-quoted strings under the function signature.

nschneid reviewed (changes requested):
Looks like a huge improvement! One thing I notice is that there are still several places elsewhere in the code where a tree is constructed with `"(ROOT "`. Would it be better to bring these under the umbrella of the new class? What about functions for token escaping?

bwaldon commented:
> One thing I notice is that there are still several places elsewhere in the code where a tree is constructed with `"(ROOT "`. Would it be better to bring these under the umbrella of the new class? What about functions for token escaping?

I totally agree. I'll bring the token escaping functions under the new class umbrella.

I think I'll also experiment with breaking up `tree_process()` -- it's doing multiple things at once right now, including canonicalization, automatic labelling, and ptree to CGEL tree conversion w/ punctuation handling... It also doesn't have a very informative name (partly because of how much it's doing). Next, I'll try to bring the functionality of `tree_process` under the umbrella of `ActivedopTree` -- I think that will let me get rid of many 'ad-hoc' tree construction lines in the code.

nschneid commented on app.py:
```diff
@@ -180,21 +177,25 @@ def __init__(self, ptree: ParentedTree, senttok: List[str], cgel_terminals = Non
 		if cgel_terminals is not None:
 			self.cgel_tree.update_terminals(cgel_terminals, gaps=True, restore_old_cat=True, restore_old_func=True)
 	
+	# brackettreestr: string representation of ptree in bracket notation, with labels consisting of a POS tag and a function tag separated by a hyphen.
```
should be in a triple-quoted string under the method name: https://peps.python.org/pep-0257/#one-line-docstrings

bwaldon replied:
Ah, sorry! Will update this right now.

bwaldon commented:
UPDATED SUMMARY: 

- Decomposes `tree_process` into multiple, more intuitively-named functions: `apply_standard_labels`, `ptree_to_cgel`, `cgel_to_canonicalized_ptree`

- Introduces the `ActivedopTree` class as a new abstraction in app.py. `ActivedopTree` class serves as a wrapper for a `ParentedTree` object, providing additional methods and attributes specific to activedop.

**Constructor** (`__init__` method):
Initializes the ActivedopTree object with a ParentedTree `ptree` and a list of tokens `senttok`. Relabels the ptree, converts to a CGEL tree, and canonicalizes the ptree with calls to `apply_standard_labels`, `ptree_to_cgel`, and `cgel_to_canonicalized_ptree`. Optionally updates the CGEL tree terminals if `cgel_terminals` are provided.

**Attributes**:
`ptree`: The ParentedTree object with numeric indices as terminals.
`senttok`: A list of tokens corresponding to the terminals of `ptree`.
`cgel_tree`: A CGELTree object.

**Methods**:
`brackettreestr`: Returns a string representation of `ptree` in bracket notation, with an option for pretty printing (which is helpful in standard, non-CGEL mode).
`treestr`: Returns a string representation of the tree (either CGEL or bracket notation, depending on app settings).
`gtree`: Returns an HTML representation of `ptree`, with an option to add an `editable` attribute to select tree nodes using  `add_editable_attribute`.
`validate`:  Returns a string, the output of validation using the standard `validate()` function (plus `validate_cgel()`, if applicable). 
`from_str` (classmethod): create an `ActivedopTree` instance from a raw string (either brackettree format or cgeltree format, depending on app settings).

nschneid commented on app.py:
```diff
@@ -153,6 +153,74 @@ def sent_escape(sent):
 
+class ActivedopTree:
+	"""Wrapper for a ParentedTree object with additional methods for activedop.
+	__init__ arguments:
```
please describe these under `__init__()`

bwaldon replied:
https://github.com/nschneid/activedop/pull/73/commits/f92e4695428e334a0f6e8d0c6bf30e10b5445df7

nschneid commented on app.py:
```diff
@@ -153,6 +153,74 @@ def sent_escape(sent):
+		else:
+			return str(self.cgel_tree)
+	
+	# gtree: 
```
delete

bwaldon replied:
https://github.com/nschneid/activedop/pull/73/commits/f92e4695428e334a0f6e8d0c6bf30e10b5445df7

nschneid commented on app.py:
```diff
@@ -544,8 +612,9 @@ def annotate(sentno):
-			tree = annotation
-			_, senttok = brackettree( "(ROOT" + cgel.parse(annotation)[0].ptb(punct = True) + ")" )
+			treeobj = ActivedopTree.from_str(annotation)
+			senttok = treeobj.senttok
```
is this being used?

bwaldon replied:
Good point -- no, there was no reason to define the `senttok` variable here. Deleted in https://github.com/nschneid/activedop/pull/73/commits/08310ce969e66f05a53402bf2dab0b554b3e83d1

nschneid commented on app.py:
```diff
@@ -897,8 +937,16 @@ def tree_process(tree : ParentedTree, senttok: List[str]) -> tuple[ParentedTree,
-		ptree_terminals.append(subt)
+	return tree_copy
 
+def ptree_to_cgel(tree: ParentedTree, senttok: List[str]) -> ParentedTree:
```
Can this be folded in as a method of `ActiveDopTree`? Or is there reason to keep it separate?

nschneid commented on app.py:
```diff
@@ -871,24 +915,20 @@ def _remove_punct(tree):
 	return number_terminals(prune_empty_non_terminals(tree_copy))
 
-def tree_process(tree : ParentedTree, senttok: List[str]) -> tuple[ParentedTree, CGELTree]:
+def apply_standard_labels(tree: ParentedTree, senttok: List[str]) -> ParentedTree:
```
Should this be a method of `ActiveDopTree`?

bwaldon replied:
Yes, I agree -- implemented in https://github.com/nschneid/activedop/pull/73/commits/655172b19c9c79a086c0776c1f6c9f0cd54229a5

nschneid commented on app.py:
```diff
@@ -153,6 +153,357 @@ def sent_escape(sent):
 	senttok = sent.split()
 	return " ".join(senttok_escape(senttok))
 
+class ActivedopTree:
```
Great to have this functionality bundled under a class. As it's getting large I would put it in a new file (activedoptree.py).

nschneid commented on app.py:
```diff
@@ -153,6 +153,357 @@ def sent_escape(sent):
+		Outputs both a CGELTree object and a ParentedTree object, the latter of which is the cleaned-up version of the input tree with a canonicalized position for punctuation preterminals/terminals.
+		"""
+		tree, senttok = self.ptree, self.senttok
+		def prune_empty_non_terminals(tree: ParentedTree) -> ParentedTree:
```
Now that we are in a class, these helper functions can be moved under the main class with underscore-prefixed names (`_prune_empty_non_terminals()`) to indicate they should only be used internally. Having them defined within a method means they will be redefined every time the method is called.

bwaldon commented:
Awesome. Here's a summary of what's in the latest commit: 

- Created a new module `activedoptree.py` which includes the `ActivedopTree` class as well as some regular expression variables and punct handling helper functions frequently invoked by the class. (Those regex variables and functions are defined outside the `ActivedopTree` class because in some cases they need to be imported in `app.py`). 
- Moved the function `workerattr` to a separate module `workerattr.py`, because it's called both within `activedoptree.py` and `app.py`. 
- Moved punctuation escaping parameters (e.g., `PUNCT_ESCAPING` and `PUNCT_TAGS`) to `settings.cfg`, because these parameters are needed both within `activedoptree.py` and `app.py`.
